src/ot_orchestration/utils/path.py

```diff
@@ -0,0 +1,345 @@
+        """Write data to Google Cloud Storage.
+
+        Args:
+            data (Any): Data to write.
+        """
+        bucket = storage.Bucket(client=self.client, name=self.bucket)
+        blob = storage.Blob(name=self.path, bucket=bucket)
+        with blob.open("w") as fp:
+            match Path(self.path).suffix:
+                case ".json":
+                    json.dump(data, fp)  # type: ignore
+                case ".yaml" | ".yml":
+                    yaml.safe_dump(data, fp, indent=2)
+                case _:
+                    fp.write(data)
+
+    def load(self) -> Any:
+        """Read data from Google Cloud Storage.
+
+        Returns:
+            Any: Loaded data.
+        """
+        bucket = storage.Bucket(client=self.client, name=self.bucket)
+        blob = storage.Blob(name=self.path, bucket=bucket)
+
+        with blob.open("r") as fp:
+            match Path(self.path).suffix:
+                case ".json":
+                    return json.load(fp)
+                case ".yaml" | ".yml":
+                    return yaml.safe_load(fp)
+                case _:
+                    return fp.read()
+
+
+class IOManager:
+    """Input Output manager class."""
+
+    def resolve(self, path: str) -> ProtoPath:
+        """Resolve path by its protocol.
+
+        Depending on the protocol the following are defined
+
+        - gs:// for Google Cloud Storage
+        - file:// for local file system
+        - no protocol for local file system
+        - s3:// (currently not implemented)
+        - s3a:// (currently not implemented)
+        - http:// (currently not implemented)
+        - https:// (currently not implemented)
+        - ftp:// (currently not implemented)
+
+        Based on registred protocols, the path is resolved to the appropriate Path object
+
+        Args:
+            path (str): Path to resolve.
+
+        Returns:
+            ProtoPath: Path object instance following ProtoPath protocol.
+        """
+        match path.split("://"):
+            case ["gs", _]:
+                return GCSPath(path)
+            case [_] | ["file", _]:
+                return NativePath(path)
+            case [protocol, _]:
+                raise NotImplementedError(
+                    "IOManager.resolve does not implement %s protocol resolution",
+                    protocol,
+                )
+            case _:
+                raise NotImplementedError(
+                    "IOManager.resolve is not implemented for path=%s", path
+                )
+
+    def resolve_paths(self, paths: list[str]) -> list[ProtoPath]:
+        """Resolve multiple paths by the protocols.
+
+        Args:
+            paths (list[str]): Paths to resolve.
+
+        Returns:
+            list[ProtoPath]: List of path objects following ProtoPath protocol.
+        """
+        return [self.resolve(p) for p in paths]
+
+    def load_many(self, paths: list[str], n_threads=100) -> list[Any]:
```

Comment:
That is a good question. I have followed [this](https://stackoverflow.com/questions/481970/how-many-threads-is-too-many.) and set up the default after some tests on ~40k files which took coulple of minutes. -  I was planning to make more tests when it comes to the processing of the real data.

Reply:
I'd say as long as it's not overly large and somehow slows down the machine, it's fine to approximate and then test as you said.

Ideally thread count should be similar to core count, as 1 thread per core gives the best performance (no overhead from switching context). But here we're dealing with heavy I/O locking ops, so that won't be the case.

Reply:
I agree, I will make a branching  for it (as testing setup will not require as many threads), after all you are right that M1 has no hypertheading :p, non the less even with it it would still be of max 16 threads.